### ISSUE TITLE
colcon-package-selection is old

### ISSUE THREAD

**Issue by kenji-miyake** [ros-tooling/setup-ros-docker#41]
## Description

Since `colcon-package-selection` in `setup-ros-docker` is old, some options like `--packages-above-and-dependencies` are not supported.

### Expected Behavior

Users can use as many options as possible that the latest version has.

### Actual Behavior

Users can't use `--packages-above-and-dependencies`.

## To Reproduce

As I've tested on CI, please see the following results.

1. Current version (0.2.6)

`osrf/ros:galactic-desktop` passed (because it's using apt-version of colcon) but `setup-ros-docker` failed.

https://github.com/kenji-miyake/setup-ros-test-colcon-version/pull/1

2. After manually updating colcon-package-selection from 0.2.6 to 0.2.10

Both `osrf/ros:galactic-desktop` and `setup-ros-docker` passed.
https://github.com/kenji-miyake/setup-ros-test-colcon-version/pull/2

## System (please complete the following information)

 - OS: Ubuntu 20.04
 - ROS 2 Distro: Galactic

**Comment by kenji-miyake** [comment]
I'll send a PR so that it will be equal to https://github.com/ros-tooling/setup-ros/pull/375.

**Comment by christophebedard** [comment]
> However, we can start by setting up `-latest` images for Rolling & Galactic to keep it simple

is this something you'd like to do @kenji-miyake? If not, I can probably do it quickly.

**Comment by kenji-miyake** [comment]
@christophebedard Hmm, I'd like to work on that, but I can't imagine the task well.
It seems the images are already named `-latest` for DockerHub and `:master` for GitHub Packages. :thinking: 
I think what we need now is to add versioned images and ask users (who want stability rather than features) to switch to the versioned images.

**Comment by kenji-miyake** [comment]
> One solution would be to add proper tests in this repo that use action-ros-ci to check that bumping something doesn't break typical workflows.

Also, I'd like to know this in a bit more detail to contribute.

Could you list up some tests? What problems do you want to prevent?
Is it only https://github.com/ros-tooling/action-ros-ci/issues/615 for now?

**Comment by christophebedard** [comment]
> It seems the images are already named `-latest` for DockerHub and `:master` for GitHub Packages. 

argh you're right! This doesn't leave us a lot of options if we want to keep similar names. I'm not sure why the name/tag naming isn't the same on Docker Hub and GitHub.

> I think what we need now is to add versioned images and ask users (who want stability rather than features) to switch to the versioned images.

I agree that we should consider this. It would be a one-time "stability break" probably only for the people using an old version of `action-ros-ci` affected by https://github.com/ros-tooling/action-ros-ci/issues/615 and it would allow us to set up a better versioning/naming system.

If we really want to avoid stability breaks, we could also abandon the `-latest`/`:master` names (i.e., leave the dependencies as-is but keep building the images periodically) and switch to another word that means the same thing to represent the real "latest" images with dependencies that are updated regularly. Maybe `main`?

..unless we swap the names by changing the Docker Hub images to `-master` and the GitHub images to `:latest` :laughing:

@emersonknapp what do you think?

> Could you list up some tests? What problems do you want to prevent?
> Is it only [ros-tooling/action-ros-ci#615](https://github.com/ros-tooling/action-ros-ci/issues/615) for now?

I don't think we should specifically try to prevent https://github.com/ros-tooling/action-ros-ci/issues/615 because it shouldn't happen anymore with newer versions of `action-ros-ci`. And I don't think we should test older versions of `action-ros-ci`, because we know it requires some older versions of `colcon-*` packages from `setup-ros*`, and if users modify their CI config I would expect them to bump all their actions/Docker images.

We can simply take & adapt some of the tests from `action-ros-ci` that represent typical use-cases relating to `colcon`/`vcs`/`rosdep`: https://github.com/ros-tooling/action-ros-ci/blob/master/.github/workflows/test.yml. That should prevent most problems.

**Comment by emersonknapp** [comment]
I agree that we should leave the existing names as-is. No breaking existing workflows.

If we're going to change names, I have some ideas about updating the scheme:

```
${OS}-${ROS_DISTRO}-${ROS_VARIANT}_v${VERSION}
```

1. For OS, skip the distro and just use ROS distro mapping
1. For [variants](https://github.com/ros2/variants) I would like to provide all 3 plus a "noinstall" option that doesn't install a variant. (open to changes on the naming). This would replace the images called e.g. `ubuntu-xenial-latest`. Also note that this means `ubuntu-foxy-noinstall` and `ubuntu-galactic-noinstall` and (for the moment right now) `ubuntu-rolling-noinstall` would be (nearly?) identical. I think this is an OK tradeoff for readability/writeability of the tags. Realistically every build job is targeting one ROS distro even if there aren't binaries preinstalled.
1. For version, I think for this type of product it makes sense to skip minor versions and just use a single increasing number. For a rolling one, i'd be fine with `devel`, `main`, `newest`, or even something like `vX` to just say it's "variable".

This would give us e.g. 
* `ubuntu-foxy-noinstall_v1`
* `ubuntu-foxy-ros-core_v1`
* `ubuntu-foxy-ros-base_v1`
* `ubuntu-foxy-desktop_v1`
* `ubuntu-galactic-noinstall_devel`
* `ubuntu-rolling-desktop_devel`

With this change, we should definitely add a README explaining how to use it, and noting the "legacy" names.

**Comment by kenji-miyake** [comment]
@emersonknapp The naming rule generally looks good to me! :+1: 
But I have some questions/suggestions.
- Are there any specific reasons for using `_` before versions instead of `-`?
- How about `bare` or `base` instead of `noinstall`?
- For version, how do you manage/express the compatibility with `setup-ros`?

**Comment by christophebedard** [comment]
>     ${OS}-${ROS_DISTRO}-${ROS_VARIANT}_v${VERSION}

That looks good!

> * How about `bare` or `base` instead of `noinstall`?

`base` might be confusing (because of the `ros-base` variant), but I think `bare` would be good.

**Comment by emersonknapp** [comment]
> How about bare or base instead of noinstall?

`bare` feels less clear to me personally - i'm still partial to `noinstall` because it means "no installed variant binaries". Maybe `novariant`? Maybe `empty`, but that sounds misleading too

> For version, how do you manage/express the correspondence with setup-ros?

I'm not sure - that's a good point. I wonder if we should also move `setup-ros` to a simple integer version too :)
